### PR TITLE
Feat/18890/button ts update

### LINKED ISSUE #18890: Migrate components to TS: Button
### Description
Migrate `Button` component to TypeScript

Dependencies
- https://github.com/MetaMask/metamask-extension/issues/19239
- #19067 
- #18889 
- #18887 
- https://github.com/MetaMask/metamask-extension/issues/18888

### Technical Details

- Convert Objects in `SNAKE_CASE` to enums in `PascalCase` (name and keys)
- Ensure enums use singular not plural e.g. `Size` not `Sizes`
- Update file name from `component-name.contants.js` to `component-name.types.ts` and add types
- Update file name from `component-name.js` to `component-name.tsx` and add types
- Update file name from `component-name.test.js` to `component-name.test.tsx`
- Update file name from `component-name.stories.js` to `component-name.stories.tsx`
- Update all instances of component objects to enums
- Update all documentation

### Acceptance Criteria

- Component has been converted to TypeScript
- Stories, documentation and tests have been updated inside of the `component-name` folder
- Before/After screen cast of component stories in storybook to show no visual regression after migration

### Difficulty: Intermediate

Good first issue for: External contributors who are familiar with TypeScript and want to contribute to improving the stability of the extension

### PR DESCRIPTION
**Pull request by garrettbear**
## Explanation
- Update `Button` unification component to TS 
- Update all objects to appropriate enums (e.g. BUTTON_VARIANTS & BUTTON_SIZES to ButtonVariant, ButtonSize)
- Small fixes to app components built in TS using `Button` and created a few lint errors (see comments) 

* Fixes #18890 

### Before

Screenshot 2023-08-21 at 3 04 15 PM
Screenshot 2023-08-21 at 3 03 52 PM
Screenshot 2023-08-21 at 3 03 15 PM

### After

Screenshot 2023-08-21 at 3 04 10 PM
Screenshot 2023-08-21 at 3 03 36 PM
Screenshot 2023-08-21 at 3 02 34 PM

## Manual Testing Steps
`yarn jest ui/components/component-library`

To ensure TS validation is working for size try to use the wrong size value on a `ButtonVariant.Primary` or `ButtonVariant.Secondary` 

Example code that should result in TS error: 
```tsx
<Button variant={ButtonVariant.Primary} size={ButtonSize.Auto}>Button Primary</Button>
```

Video example: https://www.loom.com/share/aa28283f8f5d40b8ad322f38bd7acb36

## Pre-merge author checklist

- [X] I've clearly explained:
  - [X] What problem this PR is solving
  - [X] How this problem was solved
  - [X] How reviewers can test my changes
- [ ] Sufficient automated test coverage has been added

## Pre-merge reviewer checklist

- [X] Manual testing (e.g. pull and build branch, run in browser, test code being changed)
- [X] PR is linked to the appropriate GitHub issue
- [ ] **IF** this PR fixes a bug in the release milestone, add this PR to the release milestone

If further QA is required (e.g. new feature, complex testing steps, large refactor), add the `Extension QA Board` label.

In this case, a QA Engineer approval will be be required.

### CODE REVIEW

File: ui/components/component-library/button/button.stories.tsx [hunk 1]
```diff
@@ -83,79 +63,55 @@ export default {
...
   args: {
     children: 'Button',
   },
-};
+} as Meta<typeof Button>;
 
-export const DefaultStory = (args) => <Button {...args} />;
+export const DefaultStory: StoryFn<typeof Button> = (args) => (
+  <Button {...args} />
+);
```
Comment (georgewrmarshall) [inline ui/components/component-library/button/button.stories.tsx:80]: I was wondering what the difference between creating new stories and using the `.bind()` method was. Especially when it comes to storybook. After checking ChatGPT with the question "Does the Template.bind({})  in storybook .stories.js have any performance benefits" I got:
> In summary, using Template.bind({}) can provide performance benefits in certain scenarios, especially for components with complex interactions. However, its impact might not be as significant for simpler components. As with any optimization, it's best to measure the actual impact in your specific context to determine if the performance gains justify the extra effort of using Template.bind({})


It gave me much more context so I suggest also looking it up. So although these aren't complex props I think we want to promote the convention so it's adopted widely for all stories that way if there are complex props we could get some performance benefits now that storybook is widely used.

```suggestion
const Template: StoryFn<typeof Button> = (args) => <Button {...args} />;

export const DefaultStory = Template.bind({});
```

Then for Href, Disabled, Loading, StartIconName and EndIconName you can do 
```
export const Disabled = Template.bind({});
Disabled.args = {
  disabled: true,
};

export const Loading = Template.bind({});
Loading.args = {
  loading: true,
};

export const StartIconName = Template.bind({});
StartIconName.args = {
  startIconName: IconName.AddSquare,
};

export const EndIconName = Template.bind({});
EndIconName.args = {
  endIconName: IconName.AddSquare,
};
```

File: ui/components/component-library/button/button.stories.tsx [hunk 2]
```diff
@@ -192,34 +150,38 @@ export const Block = (args) => (
   </>
 );
 
-export const As = (args) => (
-  <Box display={DISPLAY.FLEX} flexDirection={FLEX_DIRECTION.ROW} gap={2}>
+export const As: StoryFn<typeof Button> = (args) => (
```
Comment (georgewrmarshall) [inline ui/components/component-library/button/button.stories.tsx:153]: This is an existing issue but noticed on hover the href button turns black


https://github.com/MetaMask/metamask-extension/assets/8112138/43d0d1c2-d671-40e7-862d-4817d1855396

**Review by georgewrmarshall — COMMENTED** [review]
Fantastic work 💯 Everything to do with the Button looks and works great. I have a couple nits and because this PR touches a lot of files I would suggest keeping the constants file and instead of removing them in this PR deprecating them. Then in additional PRs we can go through and replace all of the enums. That way we can get this TypeScript migration PR merged and close the issue. Then we can create another issue to replace the deprecated constants and complete it in the next sprint? This will allow for easier review while we figure out the SRPQuiz type issue and this PR require a snaps devs review. It should also make it easier for them to review when the time comes


```
/**
 * @deprecated `BUTTON_VARIANT` const has been deprecated in favor of the `ButtonVariant` enum which can still be imported from `ui/components/component-library`
 */```

File: ui/components/app/srp-quiz-modal/QuizContent/QuizContent.tsx [hunk 3]
```diff
@@ -60,8 +60,7 @@ export default function QuizContent({
           key={idx}
           size={btn.size}
           onClick={btn.onClick}
-          label={btn.label}
-          variant={btn.variant}
+          variant={btn.variant as any}
```
[the remark this answers is not in the document]
Comment (georgewrmarshall) [reply ui/components/app/srp-quiz-modal/QuizContent/QuizContent.tsx:63]: Any ideas @HowardBraham?

[the remark this answers is not in the document]
Comment (DDDDDanica) [reply ui/components/app/srp-quiz-modal/QuizContent/QuizContent.tsx:63]: Hey @garrettbear i tried different approaches, but I'm confused about why we always fall in the case of `'ButtonVariant.Link | undefined`? For example, in the following warning: 
```
xxx is not assignable to type 'ButtonVariant.Link | undefined'.
```

File: ui/components/component-library/button/index.ts [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+  AUTO: Size.auto,
+};
+/**
+ * @deprecated `BUTTON_SIZES` const has been deprecated in favor of the `ButtonSize` enum which can still be imported from `ui/components/component-library`
+ */
+export const BUTTON_VARIANT = {
+  PRIMARY: 'primary',
+  SECONDARY: 'secondary',
+  LINK: 'link',
+};
```
[the remark this answers is not in the document]
Comment (georgewrmarshall) [reply ui/components/component-library/button/index.ts:23]: Go for it 😜 Saves one more place to remove I suppose 😁

**Review by georgewrmarshall — APPROVED** [review]
LGMT 🚀

**Review by darkwing — APPROVED** [review]

**Pull request merged**